packages/modules/kostal_plenticore/device.py

```diff
@@ -0,0 +1,128 @@
+from typing import Any, Callable, List, Optional
+
+from helpermodules.cli import run_using_positional_cli_args
+from modules.common import modbus
+from modules.common.abstract_device import DeviceDescriptor
+from modules.common.configurable_device import ConfigurableDevice, ComponentFactoryByType, MultiComponentUpdater
+from modules.kostal_plenticore import bat, counter, inverter
+from modules.kostal_plenticore.bat import KostalPlenticoreBat
+from modules.kostal_plenticore.config import (KostalPlenticore, KostalPlenticoreBatSetup, KostalPlenticoreCounterSetup,
+                                              KostalPlenticoreInverterSetup)
+from modules.kostal_plenticore.counter import KostalPlenticoreCounter
+from modules.kostal_plenticore.inverter import KostalPlenticoreInverter
+
+log = logging.getLogger(__name__)
+
+
+def update_components(components_todo, tcp_client: modbus.ModbusTcpClient_):
+    def little_endian_wordorder_reader(register: int, data_type: modbus.ModbusDataType):
+        return tcp_client.read_holding_registers(
+            register, data_type, unit=71, wordorder=Endian.Little)
+    reader = little_endian_wordorder_reader
+    with tcp_client:
+        for component in components_todo:
+            if isinstance(component, KostalPlenticoreBat):
+                bat_state = component.update(reader)
+        else:
+            bat_state = None
+        for component in components_todo:
+            if isinstance(component, KostalPlenticoreInverter):
+                inverter_state = component.update(reader)
+                if bat_state:
+                    dc_in = component.dc_in_string_1_2(reader)
+                    home_consumption = component.home_consumption(reader)
+                    if dc_in >= 0:
+                        if bat_state.power > 0:
+                            raw_inv_power = inverter_state.power
+                            inverter_state.power = dc_in / (dc_in + bat_state.power) * raw_inv_power
+                            bat_state.power = raw_inv_power - inverter_state.power - home_consumption
+                        else:
+                            inverter_state.power -= bat_state.power
+                component.set(inverter_state)
+            else:
+                component.update(reader)
+        if bat_state:
+            for component in components_todo:
+                if isinstance(component, KostalPlenticoreBat):
+                    component.set(bat_state)
+
+
+def create_device(device_config: KostalPlenticore):
+    tcp_client = modbus.ModbusTcpClient_(device_config.configuration.ip_address, 1502)
+    return ConfigurableDevice(
+        device_config=device_config,
+        component_factory=ComponentFactoryByType(
+            bat=bat.create_component, counter=counter.create_component, inverter=inverter.create_component),
+        component_updater=MultiComponentUpdater(
+            lambda update_components: update_components(components_todo, tcp_client)),
```

Comment:
`update_components` ist keine Funktion, sondern die Komponente. Also ein `KostalPlenticoreBat` oder `KostalPlenticoreInverter` oder `KostalPlenticoreCounter`. Diese haben praktischerweise alle eine Methode `update(self, reader: Callable[[int, ModbusDataType], Any])`. Die musst du aufrufen. So:
```python
MultiComponentUpdater(lambda component: component.update(reader))
```
Bzw. im größeren Kontext:
```python
tcp_client = ...
def little_endian_wordorder_reader(register: int, data_type: modbus.ModbusDataType):
    return tcp_client.read_holding_registers(...)

return ConfigurableDevice(
    ...
    component_updater=MultiComponentUpdater(lambda component: component.update(little_endian_wordorder_reader))
)
```

Reply:
Vor dem Aufruf der for-Schleife würde ich gerne noch den Kontextmanager für den tcp-Client setzen, damit die Verbindung nur einmal für alle Komponenten geöffnet wird. Dazu brauche ich aber den tcp_client, so wie hier: 
[https://github.com/snaptec/openWB/blob/5cc5eebb5ec631bbb32aeedc6969b4dc24c2408a/packages/modules/kostal_plenticore/device.py#L23](url)